### PR TITLE
update: mpu for t5 rpe

### PR DESCRIPTION
**Pull request by PerkzZheng**
support t5 relative position embedding mpu groups.

### CODE REVIEW

**Review by crcrpar — COMMENTED** [review]
could you add a minimal testcase to https://github.com/NVIDIA/apex/blob/master/tests/L0/run_transformer/test_parallel_state.py?

**Comment by PerkzZheng** [comment]
what kind of tests here? Following other test cases, position embeddings don't have get_rank or get_size interfaces. What we need are just get_group and is_rank_in_group.

File: apex/transformer/parallel_state.py [hunk 1]
```diff
@@ -358,6 +406,17 @@ def is_rank_in_position_embedding_group():
...
+    """Return true if current rank is in encoder relative position embedding group, False otherwise."""
+    rank = torch.distributed.get_rank()
+    global _ENCODER_RELATIVE_POSITION_EMBEDDING_GLOBAL_RANKS
+    return rank in _ENCODER_RELATIVE_POSITION_EMBEDDING_GLOBAL_RANKS
+
+def is_rank_in_decoder_relative_position_embedding_group():
+    """Return true if current rank is in decoder relative position embedding group, False otherwise."""
+    rank = torch.distributed.get_rank()
+    global _DECODER_RELATIVE_POSITION_EMBEDDING_GLOBAL_RANKS
+    return rank in _DECODER_RELATIVE_POSITION_EMBEDDING_GLOBAL_RANKS
```
Comment (crcrpar) [inline apex/transformer/parallel_state.py:419]: I think we can have a sanity check for these two functions.

Comment (PerkzZheng) [reply apex/transformer/parallel_state.py:419]: Added.

File: tests/L0/run_transformer/test_parallel_state.py [hunk 2]
```diff
@@ -120,6 +120,16 @@ def test_initialize_model_parallel_with_virtual_and_split(self) -> None:
             fake_split_rank, parallel_state.get_pipeline_model_parallel_split_rank()
         )
 
+        # relative position embedding groups check
+        self.assertEqual(
+           expected_pipeline_rank < pipeline_model_parallel_split_rank,
+           parallel_state.is_rank_in_encoder_relative_position_embedding_group();
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_parallel_state.py:126]:
```suggestion
           parallel_state.is_rank_in_encoder_relative_position_embedding_group(),
```

File: tests/L0/run_transformer/test_parallel_state.py [hunk 3]
```diff
@@ -120,6 +120,16 @@ def test_initialize_model_parallel_with_virtual_and_split(self) -> None:
...
+        )
+        self.assertEqual(
+           expected_pipeline_rank >= pipeline_model_parallel_split_rank,
+           parallel_state.is_rank_in_decoder_relative_position_embedding_group();
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_parallel_state.py:130]:
```suggestion
           parallel_state.is_rank_in_decoder_relative_position_embedding_group(),
```

**Review by crcrpar — COMMENTED** [review]
LGTM except minor comments.

**Review by crcrpar — APPROVED** [review]

**Comment by timmoon10** [comment]
I've run into errors when training GPT-3, at this line:
https://github.com/NVIDIA/apex/blob/208d9670576fe0eb411d0cecbf1ed51412524310/apex/transformer/parallel_state.py#L297

I haven't fully investigated, but I suspect `encoder_relative_position_embedding_ranks` is `[]`.

**Pull request merged**